src/cargo/core/registry.rs

```diff
@@ -335,35 +357,58 @@ http://doc.crates.io/specifying-dependencies.html#overriding-dependencies
 }
 
 impl<'cfg> Registry for PackageRegistry<'cfg> {
-    fn query(&mut self, dep: &Dependency) -> CargoResult<Vec<Summary>> {
+    fn query(&mut self,
+             dep: &Dependency,
+             f: &mut FnMut(Summary)) -> CargoResult<()> {
         // Ensure the requested source_id is loaded
         self.ensure_loaded(dep.source_id(), Kind::Normal).chain_err(|| {
             format!("failed to load source for a dependency \
                      on `{}`", dep.name())
         })?;
 
+        // Look for an override and get ready to query the real source
         let override_summary = self.query_overrides(&dep)?;
-        let real_summaries = match self.sources.get_mut(dep.source_id()) {
-            Some(src) => Some(src.query(&dep)?),
-            None => None,
+        let mut sources = self.sources.borrow_mut();
```

Comment:
Yeah I thought of originally doing that, but the call to `warn_bad_override` caught me by surprise. The mutable borrow on `self.sources` means we otherwise couldn't use `self`, so this seemed like the "cleanest" way of implementing this.

Definitely warrants a comment though!